pkg/spdk/server.go

```diff
@@ -71,59 +71,83 @@ func (s *Server) monitoringReplicas() {
 			logrus.Info("SPDK Server: stopped monitoring replicas due to the context done")
 			done = true
 		case <-ticker.C:
-			bdevLvolList, err := s.spdkClient.BdevLvolGet("", 0)
-			if err != nil {
-				logrus.Errorf("SPDK Server: failed to get lvol bdevs for replica cache update, will retry later: %v", err)
-				continue
+			if err := s.verifyReplicas; err != nil {
+				logrus.Errorf("SPDK Server: failed to verify and update replica cache, will retry later: %v", err)
 			}
-			bdevLvolMap := map[string]*spdktypes.BdevInfo{}
-			for idx := range bdevLvolList {
-				bdevLvol := &bdevLvolList[idx]
-				if len(bdevLvol.Aliases) == 0 {
-					continue
-				}
-				bdevLvolMap[spdktypes.GetLvolNameFromAlias(bdevLvol.Aliases[0])] = bdevLvol
-			}
-
-			s.Lock()
-			for _, r := range s.replicaMap {
-				r.ValidateAndUpdate(bdevLvolMap)
-			}
-			s.Unlock()
 		}
 		if done {
 			break
 		}
 	}
 }
 
+func (s *Server) verifyReplicas() error {
+	s.Lock()
+	defer s.Unlock()
+
+	bdevLvolList, err := s.spdkClient.BdevLvolGet("", 0)
+	if err != nil {
+		return err
+	}
+	subsystemList, err := s.spdkClient.NvmfGetSubsystems("", "")
+	if err != nil {
+		return err
+	}
+	bdevLvolMap := map[string]*spdktypes.BdevInfo{}
+	for idx := range bdevLvolList {
+		bdevLvol := &bdevLvolList[idx]
+		if len(bdevLvol.Aliases) == 0 {
+			continue
+		}
+		bdevLvolMap[spdktypes.GetLvolNameFromAlias(bdevLvol.Aliases[0])] = bdevLvol
+	}
+	subsystemMap := map[string]*spdktypes.NvmfSubsystem{}
+	for idx := range subsystemList {
+		subsystem := &subsystemList[idx]
+		subsystemMap[subsystem.Nqn] = subsystem
+	}
+
+	for _, r := range s.replicaMap {
+		r.ValidateAndUpdate(s.spdkClient, bdevLvolMap, subsystemMap)
+	}
+
+	return nil
+}
+
 func (s *Server) ReplicaCreate(ctx context.Context, req *spdkrpc.ReplicaCreateRequest) (ret *spdkrpc.Replica, err error) {
+	if req.Name == "" {
+		return nil, grpcstatus.Error(grpccodes.InvalidArgument, "replica name is required")
+	}
+	if req.LvsName == "" || req.LvsUuid == "" {
```

Comment:
Do we really need `LvsName`?

Reply:
Do you mean the input? We need to retain lvsName in cache `Replica` anyway (for local replica identification during engine creation). For the input, I am fine with removing it. But we need to get the lvs from SPDK if the name not provided.

Reply:
I see. We can keep it here.